### PR TITLE
feat(provider/kubernetes): Cache arbitrary resource

### PR DESCRIPTION
**Pull request by lwander**
This starts only with the replica set, but will make adding new caching agents much easier.

### CODE REVIEW

**Comment by lwander** [comment]
spinnaker/spinnaker#1902
spinnaker/spinnaker#1919

**Comment by lwander** [comment]
paging the @duftler

File: clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesSelectorSpec.groovy [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package com.netflix.spinnaker.clouddriver.kubernetes.v2.security
+
+import spock.lang.Specification
+import spock.lang.Unroll
+
+class KubernetesSelectorSpec extends Specification {
+  @Unroll
+  void "correctly turns the desired selector in to a k8s selector string (#result)"() {
```
Comment (duftler) [inline clouddriver-kubernetes/src/test/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesSelectorSpec.groovy]: s/in to/into

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/Keys.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.KubernetesApiVersion;
+import com.netflix.spinnaker.clouddriver.kubernetes.v2.description.KubernetesKind;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class Keys {
+  public enum Kind {
+    LOGICAL("logical"),
```
[the remark this answers is not in the document]
Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/Keys.java]: I just preferred the `.name` access, similar to how we did `.ns` on the keys for all the other providers, but now that you point it out it doesn't feel very  java-like.

[the remark this answers is not in the document]
Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/Keys.java]: SGTM

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCacheDataConverter.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+@Slf4j
+public class KubernetesCacheDataConverter {
+  public static CacheData fromResource(String account, ObjectMapper mapper, Object resource) {
+    KubernetesManifest manifest = mapper.convertValue(resource, KubernetesManifest.class);
+    KubernetesKind kind = manifest.getKind();
+    KubernetesApiVersion apiVersion = manifest.getApiVersion();
+    String name = manifest.getName();
+    String namespace = manifest.getNamespace();
+
+    Map<String, Object> attributes = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/caching/agent/KubernetesCacheDataConverter.java]: Ah I was looking for something like that.

**Pull request merged**